modules/core/field/src/Plugin/Field/FieldFormatter/UrlLinkFormatter.php

```diff
@@ -30,17 +30,17 @@ public function viewElements(FieldItemListInterface $items, $langcode) {
     // a link if the URI is a valid URL.
     foreach ($items as $delta => $item) {
```

Comment:
```suggestion
    /** @var \Drupal\Core\Field\Plugin\Field\FieldType\UriItem $item */
    foreach ($items as $delta => $item) {
```

This might be a better solution. I believe we can provide a better type `[Field]Item` hint for `$item`.

The change in phpstan-drupal added a stub for the `FormatterInterface::viewElements()` function which specifies that each is item is a generic `FieldItemInterface`, which has no `$value` property definition: https://github.com/mglaman/phpstan-drupal/pull/822

But phptan-drupal provides stub files for nearly all of the core field types to add correct property annotations. Nearly all of these specify the `$value` property: https://github.com/mglaman/phpstan-drupal/tree/main/stubs/Drupal/Core/Field/Plugin/Field/FieldType

Reply:
Ah nice thanks for the explanation @paul121!

I tested the `/** @var ...` line locally and it works! The only one that is still affected is `InventoryFormatter`, which is for our own custom `InventoryItem` field type... so `phpstan-drupal` doesn't provide a stub for that one.

Adding ` * @property string $value` to our `InventoryItem` docblock comment fixes it.

I pushed a fixup commit demonstrating this. In undoes the changes to `->value` and adds the type comments instead. What do you think? (cc @wotnak)

https://github.com/farmOS/farmOS/pull/943/commits/7499b4cde7b446e3154ecdbda246f9cb22826e72

Reply:
Looks ok.

For consistency, might be good to also declare types for other InventoryItem's properties (`measure`, `units`).

For narrowing types, I usually prefer to use `instanceof` since it actually checks the type while running the code instead of only assuming one for type checker tools. So here instead of `@var` comment we could use `if (!$item instanceof UriItem) { continue; }` inside the foreach.

Reply:
Ah interesting @wotnak ! I had also added some `assert()` checks in tests, but feel like `instanceof()` is better for runtime code. Curious if you have any other insights on runtime assertions vs instanceof.

https://github.com/farmOS/farmOS/pull/906#discussion_r1923073105

Reply:
@paul121 Assertions are fine for tests, but for actual application code it's usually better to use other ways of checking values since assert() by default throws an error when the assertion is not upheld and, depending on [php configuration](https://www.php.net/manual/en/ini.core.php#ini.zend.assertions), they can even be completely ignored at runtime.